exercises/react/react.ts

```diff
@@ -0,0 +1,102 @@
+//
+// Delete and replace stub with your own implementation
+//
+// Inspired by "How it Works":
+// https://indepth.dev/posts/1269/finding-fine-grained-reactive-programming#how-it-works
+// https://levelup.gitconnected.com/finding-fine-grained-reactive-programming-89741994ddee?source=friends_link&sk=31c66a70c1dce7dd5f3f4229423ad127#4543
+//
+// and "Computations":
+// https://github.com/ryansolid/solid/blob/master/documentation/reactivity.md#user-content-computations
+//
+
+type EqualFn<T> = (lhs: T, rhs: T) => boolean
+type GetterFn<T> = () => T
+type SetterFn<T> = (v: T) => T
+type UnsubscribeFn = () => void
+type UpdateFn<T> = (value?: T) => T
+
+type InputCouple<T> = [GetterFn<T>, SetterFn<T>]
+
+type Options = {
+  name: string // for debugging
+}
+
+type ObserverR = {
+  name?: string
+}
+
+type ObserverV<T> = {
+  value?: T
+  fn: UpdateFn<T>
+}
+
+type Observer<T> = ObserverR & ObserverV<T>
+
+type SubjectR = {
+  name?: string
+  observer: ObserverR | undefined
+}
+
+type SubjectV<T> = {
+  value: T
+  equal?: EqualFn<T>
+}
+
+type Subject<T> = SubjectR & SubjectV<T>
+
+// module Context value
+let activeObserver: ObserverR
+
+function updateObserver<T>(observer: Observer<T>): void {
+  const prevObserver = activeObserver
+  activeObserver = observer
+  observer.value = observer.fn(observer.value)
+  activeObserver = prevObserver
+}
+
+function createInput<T>(
+  value: T,
+  _equal?: boolean | EqualFn<T>,
+  options?: { name?: string }
+): InputCouple<T> {
+  const s: Subject<T> = {
+    name: options?.name,
+    observer: undefined,
+    value,
+    equal: undefined,
+  }
+
+  const read: GetterFn<T> = () => {
+    if (activeObserver) s.observer = activeObserver
+    return s.value
+  }
+
+  const write: SetterFn<T> = (value) => {
+    s.value = value
+    if (s.observer) updateObserver(s.observer as Observer<unknown>)
+    return s.value
+  }
+
+  return [read, write]
+}
+
+function createComputed<T>(
+  fn: UpdateFn<T>,
+  value?: T,
+  _equal?: boolean | EqualFn<T>,
+  options?: { name?: string }
+): GetterFn<T> {
+  const o: Observer<T> = {
+    name: options?.name,
+    value,
+    fn,
+  }
+  updateObserver(o)
+  return (): T => o.value!
+}
+
+function createCallback<T>(_fn: UpdateFn<T>, _value?: T): UnsubscribeFn {
+  return (): void => {} // eslint-disable-line @typescript-eslint/no-empty-function
```

Comment:
```suggestion
  return (): void => {}
```

Okay to fail for stubs

Reply:
```
99:22  error  Unexpected empty arrow function  @typescript-eslint/no-empty-function
```

so I changed it to
 
```typescript
/**
 * Creates a callback closure with the passed function
 * which is expected to perform side effects.
 *
 * @privateRemarks
 * `observer` isn't mean't to be an empty object literal.
 * Replace it with something more appropriate to its
 * purpose.
 *
 * @param fn    - Callback function. This function
 *              references one or more accessors of
 *              subjects. It may perform side effects.
 *              It will also be passed the
 *              value that it returned the last time it
 *              was invoked.
 * @param value - Initial value that is passed to `fn` when
 *              when it executes for the first time.
 * @returns     - An `unsubscribe` function. Once invoked
 *              the callback closure will stop receiving
 *              updates from the subjects it originally
 *              subscribed to.
 */
function createCallback<T>(_fn: UpdateFn<T>, _value?: T): UnsubscribeFn {
  const observer = {};
  return ((observer: unknown | undefined) => (): void => {
    if (!observer) return
    observer = undefined;
    // i.e. dispose of any active subscriptions
  })(observer)
}
```

* It's a big nudge to implement a separate `makeUnsubscribe` function
* Lets us keep the `const observer` without going to `let observer`
* `observer: unknown | undefined` is more appropriate because we are not communicating optionality like `observer?: unknown` does - we are reserving the right to set it to `undefined` in the returned function.

Reply:
I'll fix `99:22  error  Unexpected empty arrow function  @typescript-eslint/no-empty-function` upstream. It shouldn't complain about stubs 👍🏽